src/DataReader/proto_writer.rs

```diff
@@ -0,0 +1,334 @@
+) -> ProtoDecisionPoint {
+    let source = state_to_proto_state(decision_point.source(), system);
+
+    let edges = decision_point
+        .possible_decisions()
+        .iter()
+        .map(edge_id_to_proto_edge)
+        .collect();
+
+    ProtoDecisionPoint {
+        source: Some(source),
+        edges,
+    }
+}
+
+fn state_to_proto_state(s: &State, system: &TransitionSystemPtr) -> ProtoState {
+    let location_tuple = location_tuple_to_proto_location_tuple(s.get_location());
+    let federation = federation_to_proto_federation(s.zone_ref(), system);
+
+    ProtoState {
+        location_tuple: Some(location_tuple),
+        federation: Some(federation),
+    }
+}
+
+fn location_tuple_to_proto_location_tuple(l: &LocationTuple) -> ProtoLocationTuple {
+    ProtoLocationTuple {
+        locations: location_id_to_proto_location_vec(&l.id),
+    }
+}
+
+fn location_id_to_proto_location_vec(id: &LocationID) -> Vec<ProtoLocation> {
+    match id {
+        LocationID::Simple {
+            location_id,
+            component_id,
+        } => vec![ProtoLocation {
+            id: location_id.to_string(),
+            specific_component: Some(SpecificComponent {
+                component_name: component_id.as_ref().unwrap_or(&"".to_string()).to_string(), // TODO this looks disgusting
+                component_index: 0,
+            }),
+        }],
+        LocationID::Conjunction(l, r)
+        | LocationID::Composition(l, r)
+        | LocationID::Quotient(l, r) => location_id_to_proto_location_vec(l)
+            .into_iter()
+            .chain(location_id_to_proto_location_vec(r).into_iter())
+            .collect(),
+        LocationID::AnyLocation() => vec![],
+    }
+}
+
+fn federation_to_proto_federation(
+    federation: &OwnedFederation,
+    system: &TransitionSystemPtr,
+) -> ProtoFederation {
+    ProtoFederation {
+        disjunction: Some(disjunction_to_proto_disjunction(
+            &federation.minimal_constraints(),
+            system,
+        )),
+    }
+}
+
+fn disjunction_to_proto_disjunction(
+    disjunction: &Disjunction,
+    system: &TransitionSystemPtr,
+) -> ProtoDisjunction {
+    ProtoDisjunction {
+        conjunctions: disjunction
+            .conjunctions
+            .iter()
+            .map(|conjunction| conjunction_to_proto_conjunction(conjunction, system))
+            .collect(),
+    }
+}
+
+fn conjunction_to_proto_conjunction(
+    conjunction: &Conjunction,
+    system: &TransitionSystemPtr,
+) -> ProtoConjunction {
+    ProtoConjunction {
+        constraints: conjunction
+            .constraints
+            .iter()
+            .map(|constraint| constraint_to_proto_constraint(constraint, system))
+            .collect(),
+    }
+}
+
+fn constraint_to_proto_constraint(
+    constraint: &Constraint,
+    system: &TransitionSystemPtr,
+) -> ProtoConstraint {
+    fn clock_name(clock_name_and_component: Option<&(String, String)>) -> String {
+        let ZERO_CLOCK_NAME = "0";
```

Comment:
can't you use the const keyword here?

Reply:
Also this should be a string as you are calling to_string anyway

Reply:
> can't you use the const keyword here?
Yes, we will do that.

> Also this should be a string as you are calling to_string anyway
It's not possible to call non-const functions in const declaration.